src/utils/vectorEncoder.ts

```diff
@@ -0,0 +1,398 @@
+ * encoding option
+ * @param {Array} data
+ * @param {number} firstX
+ * @param {number} intervalX
+ * @param {string} encoding: ('FIX','SQZ','DIF','DIFDUP','CVS','PAC') Default 'DIFDUP'
+ * @return {string}
+ */
+export function encode(
+  data: DoubleArray,
+  firstX: number,
+  intervalX: number,
+  encoding: string,
+) {
+  switch (encoding) {
+    case 'FIX':
+      return fixEncoding(data, firstX, intervalX);
+    case 'SQZ':
+      return squeezedEncoding(data, firstX, intervalX);
+    case 'DIF':
+      return differenceEncoding(data, firstX, intervalX);
+    case 'DIFDUP':
+      return differenceDuplicateEncoding(data, firstX, intervalX);
+    case 'CSV':
+      return commaSeparatedValuesEncoding(data, firstX, intervalX);
+    case 'PAC':
+      return packedEncoding(data, firstX, intervalX);
+    default:
+      return differenceEncoding(data, firstX, intervalX);
+  }
+}
+
+/**
+ * @private
+ * No data compression used. The data is separated by a comma(',').
+ * @param data
+ * @param firstX
+ * @param intervalX
+ * @return {string}
+ */
+export function commaSeparatedValuesEncoding(
+  data: DoubleArray,
+  firstX: number,
+  intervalX: number,
+) {
+  return fixEncoding(data, firstX, intervalX, ',');
+}
+
+/**
+ * @private
+ * No data compression used. The data is separated by the specified separator.
+ * @param  data
+ * @param firstX
+ * @param intervalX
+ * @param  separator, The separator character
+ * @return {string}
+ */
+export function fixEncoding(
+  data: DoubleArray,
+  firstX: number,
+  intervalX: number,
+  separator = ' ',
+) {
+  let outputData = '';
+  let j = 0;
+  let TD = data.length;
+  let i;
+  while (j < TD - 7) {
+    outputData += Math.ceil(firstX + j * intervalX);
+    for (i = 0; i < 8; i++) {
+      outputData += separator + data[j++];
+    }
+    outputData += newLine;
+  }
+  if (j < TD) {
+    // We add last numbers
+    outputData += Math.ceil(firstX + j * intervalX);
+    for (i = j; i < TD; i++) {
+      outputData += separator + data[i];
+    }
+  }
+  return outputData;
+}
+
+/**
+ * @private
+ * No data compression used. The data is separated by the sign of the number.
+ * @param {Array} data
+ * @param {number} firstX
+ * @param {number} intervalX
+ * @return {string}
+ */
+export function packedEncoding(
+  data: DoubleArray,
+  firstX: number,
+  intervalX: number,
+) {
+  let outputData = '';
+  let j = 0;
+  let TD = data.length;
+  let i;
+
+  while (j < TD - 7) {
+    outputData += Math.ceil(firstX + j * intervalX);
+    for (i = 0; i < 8; i++) {
+      if (data[j] < 0) {
+        outputData += `-${data[j++]}`;
+      } else {
+        outputData += `+${data[j++]}`;
+      }
+    }
+    outputData += newLine;
+  }
+  if (j < TD) {
+    // We add last numbers
+    outputData += Math.ceil(firstX + j * intervalX);
+    for (i = j; i < TD; i++) {
+      if (data[i] < 0) {
+        outputData += `-${data[i]}`;
```

Comment:
I don't think this is correct. Is there any testcase ? You will have 2 times '-'. Only in the case of positive numbers you need to add the sign.